src/sfizz/VoiceStealing.h

```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: BSD-2-Clause
+
+// This code is part of the sfizz library and is licensed under a BSD 2-clause
+// license. You should have receive a LICENSE.md file along with the code.
+// If not, contact the sfizz maintainers at https://github.com/sfztools/sfizz
+
+#pragma once
+
+#include "Config.h"
+#include "Voice.h"
+#include "SisterVoiceRing.h"
+#include <vector>
+#include "absl/types/span.h"
+
+namespace sfz
+{
+class VoiceStealing
+{
+public:
+    VoiceStealing()
+    {
+        voiceScores.reserve(config::maxVoices);
+    }
+
+    Voice* steal(absl::Span<Voice*> voices) noexcept
+    {
+        // Start of the voice stealing algorithm
+        absl::c_sort(voices, voiceOrdering);
+
+        const auto sumEnvelope = absl::c_accumulate(voices, 0.0f, [](float sum, const Voice* v) {
+            return sum + v->getAverageEnvelope();
+        });
+        const auto envThreshold = sumEnvelope
+            / static_cast<float>(voices.size()) * config::stealingEnvelopeCoeff;
+        const auto ageThreshold = voices.front()->getAge() * config::stealingAgeCoeff;
```

Comment:
Is the span of voices assumed to be sorted by age?
Isn't the `swapAndPop` idion at risk to mess with this order?

Btw unrelated, regarding `swapAndPopAll`, since this thing is O(N) anyway, you may as well replace it with order-preserving removal for identical complexity.

Reply:
Yes it would but I sort it at the beginning of the method, so it's fine here.

Not sure I get your point about swapAndPopAll?

Reply:
I don't swap in here.

Reply:
This statement makes an assumption on the ordering of voices as they are traversed, no ?
`if (ref->getAge() < ageThreshold)`
I don't really understand what this is trying to express.

>Not sure I get your point about swapAndPopAll?

I'll make it clear with an example.

Here below, the `vector::erase` provides a stable sort, but swap-and-pop does not. (1-2-3 vs 1-3-2)
The runtime complexity is exactly the same.

```
    std::vector<int> v;
    auto predicate = [](int x) { return x > 3; };

    // using swap & pop method
    v = { 1, 4, 2, 5, 3 };
    sfz::swapAndPopAll(v, predicate);
    for (int x : v) std::cout << " " << x; std::cout << "\n";

    // using vector::erase method
    v = { 1, 4, 2, 5, 3 };
    auto it = std::remove_if(v.begin(), v.end(), predicate);
    v.resize(std::distance(v.begin(), it));
    for (int x : v) std::cout << " " << x; std::cout << "\n";
```

Reply:
> This statement makes an assumption on the ordering of voices as they are traversed, no ?
> `if (ref->getAge() < ageThreshold)`
> I don't really understand what this is trying to express.

Yes, but the voices are sorted at the beginning of the method (`absl::c_sort(voices, voiceOrdering);`). The `voiceOrdering` predicate sorts by age. So it's ok to assume this order here. 

This condition `ref->getAge() < ageThreshold` avoids killing voices that are too young compared to the rest. I'll check if it's entirely needed in the newer versions, because this was in early in my attempts to improve this algorithm.

> I'll make it clear with an example.
>
> Here below, the vector::erase provides a stable sort, but swap-and-pop does not. (1-2-3 vs 1-3-2)
The runtime complexity is exactly the same.

Just to be sure, this discussion is indeed unrelated to the actual stealing method?

The complexity in terms of comparisons is the same but the number of swaps would be higher with remove_if, unless I'm missing something? The remove_if version *could* have the same number of swaps if it goes through the vector and then resolves the swaps to do after, but i) this is an effect of having only 5 elements in the example, and ii) can we count on remove_if implementations to operate like this?

Reply:
>Yes, but the voices are sorted at the beginning of the method (absl::c_sort(voices, voiceOrdering);). The voiceOrdering predicate sorts by age. So it's ok to assume this order here.

Ok I see. Maybe it makes sense to make use of a policy pattern for these?
I recall this pattern to be present in some places in boost libraries.

I don't find where this is, but something along the lines of
`void search(T x, already_sorted_t) // the fast implementation`
`void search(T x, not_already_sorted_t = {}) // the slower one`

>Just to be sure, this discussion is indeed unrelated to the actual stealing method?

Yes it's unrelated